### PR TITLE
[somfytahoma] bug fixes for the V2 API

### PR DESCRIPTION
**Pull request by octa22**
This PR contains bug fixes for the API v2 migration mentined in the forum
https://community.openhab.org/t/tahoma-binding-compatible-with-oh2/27004/388

Fixes https://github.com/openhab/openhab-addons/issues/6533

changes:
- bug fixes (SmokeSensor, Light, "too many requests")
- improved the device discovery
- improved MY position setting
- added the cloud reconciliation feature (regular sync against the Tahoma Cloud)
- added support for the Water & Humidity sensor devices
- added support fot the Myfox camera & alarm devices
- added support for the Roller shutter UNO devices

### CODE REVIEW

**Comment by wborn** [comment]
I've changed the base branch of this PR to 2.5.x @octa22, see https://github.com/openhab/openhab-addons/issues/6702.

**Comment by martinvw** [comment]
It seems that some conflict was introduced, it would be great if you could rebase, thanks!

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaMyfoxCameraHandler.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        super.handleCommand(channelUID, command);
+        if (!SHUTTER.equals(channelUID.getId())) {
+            return;
+        }
+
+        if (RefreshType.REFRESH.equals(command)) {
```
Comment (martinvw) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaMyfoxCameraHandler.java]: Could be an ==

**Comment by octa22** [comment]
> Do all people require this now or can people still be using/needing the old API? It looks good to me.

The Somfy v2 API is already part of the OH 2.5.0, this PR contains some bug fixes and adds support for some new things. Hope it will be part of the next release (affected users may get the latest version in the forum)

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 2]
```diff
@@ -396,6 +408,34 @@ private void stopPolling() {
         return null;
     }
 
+    public @Nullable List<SomfyTahomaDevice> getDevices() {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java]: Would it be possible to return `Collections.emptyList()`? Then you don't need to do a null-check on the result

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 3]
```diff
@@ -497,32 +531,40 @@ private void processGatewayEvent(SomfyTahomaEvent event) {
...
-                Thing th = getThingByDeviceUrl(url);
-                if (th == null) {
-                    continue;
-                }
-                SomfyTahomaBaseThingHandler handler = (SomfyTahomaBaseThingHandler) th.getHandler();
-                if (handler != null) {
-                    handler.updateThingStatus(states);
-                    handler.updateThingChannels(states);
-                }
+        List<SomfyTahomaDevice> devices = getDevices();
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java]: If you return the empty list above:

```
getDevices().forEach(device -> updateDevice(device));
```

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaGatewayHandler.java [hunk 4]
```diff
@@ -39,21 +40,20 @@ public SomfyTahomaGatewayHandler(Thing thing) {
 
     @Override
     public void initialize() {
-        if (getBridge().getStatus() == ThingStatus.ONLINE) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaGatewayHandler.java]: I think you don't need to override if you do nothing but calling super.

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaMyfoxCameraHandler.java (hunk 1 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaMyfoxCameraHandler.java]: or `instanceof RefreshType`

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java [hunk 5]
```diff
@@ -25,19 +25,28 @@
 public class SomfyTahomaDeviceDefinition {
 
     private ArrayList<SomfyTahomaDeviceDefinitionCommand> commands = new ArrayList<>();
+    private ArrayList<SomfyTahomaDeviceDefinitionState> states = new ArrayList<>();
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java]:
```suggestion
    private List<SomfyTahomaDeviceDefinitionState> states = new ArrayList<>();
```

also in the line above

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java [hunk 6]
```diff
@@ -25,19 +25,28 @@
...
 
     public ArrayList<SomfyTahomaDeviceDefinitionCommand> getCommands() {
         return commands;
     }
 
+    public ArrayList<SomfyTahomaDeviceDefinitionState> getStates() {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java]:
```suggestion
    public List<SomfyTahomaDeviceDefinitionState> getStates() {
```

also above. or do you require the list to be an `ArrayList`?

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinitionState.java [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+ * The {@link SomfyTahomaDeviceDefinitionState} holds information about states
+ * provided by a device.
+ *
+ * @author Ondrej Pecta - Initial contribution
+ */
+@NonNullByDefault
+public class SomfyTahomaDeviceDefinitionState {
+
+    private String qualifiedName = "";
+    private ArrayList<String> values = new ArrayList<>();
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinitionState.java]: see above

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaMyfoxCameraHandler.java (hunk 1 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaMyfoxCameraHandler.java]: changed to the latter one

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/discovery/SomfyTahomaItemDiscoveryService.java [hunk 8]
```diff
@@ -261,27 +303,32 @@ private boolean isStateLess(SomfyTahomaDevice device) {
...
             }
-            logger.info("Device states: {}", sb);
+            logger.info("Current device states: {}", sb);
         }
     }
 
     private boolean hasState(SomfyTahomaDevice device, String state) {
-        for (SomfyTahomaState st : device.getStates()) {
-            if (state.equals(st.getName())) {
+        SomfyTahomaDeviceDefinition def = device.getDefinition();
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/discovery/SomfyTahomaItemDiscoveryService.java]: ok

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBaseThingHandler.java [hunk 9]
```diff
@@ -247,20 +273,24 @@ private void updateThingStatus(@Nullable SomfyTahomaState state) {
...
-                        }
-                    }
+            properties.put(state.getName(), state.getValue().toString());
+            updateThingChannels(state);
+        }
+        updateProperties(properties);
+    }
+
+    public void updateThingChannels(SomfyTahomaState state) {
+        for (HashMap.Entry<String, String> entry : stateNames.entrySet()) {
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBaseThingHandler.java]: ok

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java (hunk 2 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java]: ok

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java (hunk 3 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java]: done

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaGatewayHandler.java (hunk 4 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaGatewayHandler.java]: OK

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java (hunk 5 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java]: ok

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java (hunk 6 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinition.java]: OK

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinitionState.java (hunk 7 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaDeviceDefinitionState.java]: ok

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaAction.java [hunk 10]
```diff
@@ -26,7 +27,7 @@
 public class SomfyTahomaAction {
 
     private String deviceURL = "";
-    private ArrayList<SomfyTahomaCommand> commands = new ArrayList<SomfyTahomaCommand>();
+    private List<SomfyTahomaCommand> commands = new ArrayList<SomfyTahomaCommand>();
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaAction.java]: Sorry, didn't see that. Use diamond operator where possible

```
     private List<SomfyTahomaCommand> commands = new ArrayList<>();
```

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaCommand.java [hunk 11]
```diff
@@ -27,7 +28,7 @@
 
     int type;
     String name;
-    ArrayList<Object> parameters = new ArrayList<Object>();
+    List<Object> parameters = new ArrayList<Object>();
```
Comment (J-N-K) [inline bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaCommand.java]:
```
 List<Object> parameters = new ArrayList<>();
```

That is out of scope of this PR, but: Is it really necessary that this is `Object`? In `getPercentParameters` it is assumed that all list-members can be casted to `Double`.

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaAction.java (hunk 10 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaAction.java]: I have got rid of the obsolete SomfyTahomaCommand. thanks

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaCommand.java (hunk 11 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/model/SomfyTahomaCommand.java]: I have got rid of the obsolete SomfyTahomaCommand. thanks

**Comment by J-N-K** [comment]
Thanks.

**Pull request merged**